### PR TITLE
Add generic Set type and use it

### PR DESCRIPTION
**Pull request by Porges**
Based upon expanding the `StringSet` type. Note that we cannot substitute uses of `map[T]struct{}` where `T` is an interface because interface types are [not comparable](https://github.com/golang/go/issues/51338), even though the `map[T]struct{}` version works.

`AreEqual` is a standalone function at the moment because making it a method [crashes the compiler](https://github.com/golang/go/issues/51840).

### CODE REVIEW

File: v2/tools/generator/internal/codegen/pipeline/recursivetypefixer/simple_recursive_type_fixer.go [hunk 1]
```diff
@@ -68,16 +69,14 @@ func (s *SimpleRecursiveTypeFixer) unrollObjectTypeProperty(ot *astmodel.ObjectT
 	// quick scan for others. This is to deal with situations where there are multiple properties we need to remove.
 	// Without this check, we end up generating N different types (each with 1 property removed) which isn't allowed
 	// because types with the same name must match structurally when we go to add them to the type collection later.
-	toRemove := map[astmodel.PropertyName]struct{}{
-		prop.PropertyName(): {},
```
Comment (Porges) [inline v2/tools/generator/internal/codegen/pipeline/recursivetypefixer/simple_recursive_type_fixer.go:72]: this right here is one of the worse syntax trainwrecks i've seen 😆

File: v2/tools/generator/internal/codegen/pipeline/report_resource_versions.go [hunk 2]
```diff
@@ -92,26 +93,25 @@ func (report *ResourceVersionsReport) WriteTo(outputPath string, samplesURL stri
...
-		err = os.MkdirAll(outputPath, 0700)
+		err = os.MkdirAll(outputPath, 0o700)
 		if err != nil {
 			return errors.Wrapf(err, "Unable to create directory %q", outputPath)
 		}
 	}
 
 	destination := path.Join(outputPath, "resources.md")
-	return ioutil.WriteFile(destination, []byte(buffer.String()), 0600)
+	return ioutil.WriteFile(destination, []byte(buffer.String()), 0o600)
```
Comment (theunrepentantgeek) [inline v2/tools/generator/internal/codegen/pipeline/report_resource_versions.go:103]: Is the leading `0` not sufficient to specify Octal?

**Review by theunrepentantgeek — APPROVED** [review]

**Review by theunrepentantgeek — APPROVED** [review]

**Review by matthchr — APPROVED** [review]
Approved with some comments

File: v2/tools/generator/internal/codegen/pipeline/report_resource_versions.go (hunk 2 above)
Comment (Porges) [reply v2/tools/generator/internal/codegen/pipeline/report_resource_versions.go:103]: auto-formatter did this (to make it clearer, leading `0` is not obvious!)

File: v2/internal/set/set.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"golang.org/x/exp/slices"
+)
+
+type StringSet = Set[string]
+
+// Set provides a standard way to have a set of distinct things
+type Set[t comparable] map[t]struct{}
+
+// MakeStringSet creates a new set with the supplied strings
+func MakeStringSet(strings ...string) StringSet {
```
[the remark this answers is not in the document]
Comment (Porges) [reply v2/internal/set/set.go:20]: yes, to minimize changes

File: v2/internal/set/set.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	return maps.Equal(left, right)
+}
+
+// Values returns a slice of all the values in the set
+func (set Set[t]) Values() []t {
+	return maps.Keys(set)
+}
+
+// AsSortedSlice returns a sorted slice of values from this set
+func AsSortedSlice[t constraints.Ordered](set Set[t]) []t {
```
[the remark this answers is not in the document]
Comment (Porges) [reply v2/internal/set/set.go:76]: yes, you can’t add constraints in a method

File: v2/internal/set/set.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Porges) [reply v2/internal/set/set.go:20]: decided to just update all usages, there weren’t that many

File: v2/internal/set/set.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import (
+	"golang.org/x/exp/constraints"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+type StringSet = Set[string]
+
+// Set provides a standard way to have a set of distinct things
+type Set[t comparable] map[t]struct{}
```
[the remark this answers is not in the document]
Comment (Porges) [reply v2/internal/set/set.go]: had a look at the examples and they use uppercase, so switched

**Pull request merged**